_posts/2020-06-05-current-state-of-distributed-dask-clusters.md

```diff
@@ -0,0 +1,338 @@
+
+The mechanisms for starting and stopping workers are added via plugins. Many of the implementations we are about to discuss include this logic.
+
+## Connectivity
+
+Dask uses TCP to communicate between client, scheduler and workers by default. This means that all of these components must be on a TCP/IP network with open routes between the machines. Many connectivity problems step from firewalls or private networks blocking connections between certain components. An example of this would be running Dask on a cloud platform like AWS, but running the Python session and client on your laptop while sitting in a coffee shop using the free wifi. You must ensure you are able to route traffic between components, either by exposing the Dask cluster to the internet or by connecting your laptop to the private network via a VPN or tunnel.
+
+There is also [ongoing work to add support for UCX](https://blog.dask.org/2019/06/09/ucx-dgx) to Dask, which will allow it to make use of InfiniBand or NVLink networks where they are available.
+
+## Cluster Managers
+
+In the following section we are going to cover a range of cluster manager implementations which are available within the Dask community.
+
+In the Dask distributed codebase there is a `Cluster` superclass which can be subclassed to build various cluster managers for different platforms. Members of the community have taken this and built their own packages which enable creating a Dask cluster on a specific platform, for example [Kubernetes](https://kubernetes.io/).
+
+The design of these classes is that you import the cluster manager into your Python session and instantiate it. The object then handles starting the scheduler and worker processes on the target platform. You can then create a `Client` object as usual from that cluster object to connect to it.
+
+All of these cluster manager objects are ephemeral clusters, they only exist for the duration of the Python session and then will be cleaned up.
+
+### Local Cluster
+
+Let's start with one of the reference implementations of `Cluster` from the Dask distributed codebase `LocalCluster`.
+
+```python
+from dask.distributed import LocalCluster, Client
+
+cluster = LocalCluster()
+client = Client(cluster)
+```
+
+This cluster manager starts a scheduler on your local machine, and then starts a worker for every CPU core that it finds on the machine.
+
+### SSH Cluster
+
+Another reference implementation is `SSHCluster`. This is one of the most pure and simple ways of using multiple machines with Dask distributed and is very similar to our initial example in this blog post.
+
+```python
+from dask.distributed import SSHCluster, Client
+
+cluster = SSHCluster(["MachineA", "MachineB", "MachineC"])
```

Comment:
How to do I configure `SSHCluster` to properly authenticate on each machine? I'm not seeing any information on how to do that in [the docs][1].

[1]:https://docs.dask.org/en/latest/setup/ssh.html#distributed.deploy.ssh.SSHCluster

Reply:
Good point!

General best practice would be to use your `~/.ssh/config` file to set any options.

However, if you want to specify in your code you pass the `connect_options` kwarg with a dictionary of options for [asyncssh.connect](https://asyncssh.readthedocs.io/en/latest/api.html#connect).

Also as of the latest release, you can pass a list of dicts to `connect_options` to specify bespoke config for each host.

Do you have any suggestions on how we could update the docs/blog post to make this clearer?

Reply:
As a start I've opened dask/distributed#3859.

Reply:
Thanks for the clarification! I ask because I'm essentially replicating the functionality of `SSHCluster` with my current setup. I launch a Dask scheduler/workers on my cluster, then use port forwarding to connect Dask client to the scheduler. That way, I can develop on a Jupyter notebook on my own machine and task all the computation to my cluster.

 In this post, it might be useful to have a snippet from the "best practices" that covers the basic use case, when the user can SSH into a machine and enters their password after typing `ssh user@machineA`. If I were writing this post, I'd link to the `SSHCluster` docstring for interested users.

I'll comment in dask/distributed#3859 for some more details on the docstring.

Reply:
>  If I were writing this post, I'd link to the SSHCluster docstring for interested users.

Agreed!